File: data-structure/java/CircularQueue.java
public class CircularQueue {
    private int capacity;
    private int values[];
    private int first = 0;
    private int last = -1;
    private int elemNumber = 0;

    public CircularQueue(int capacity) {
        this.capacity = capacity;
        this.values = new int[capacity];
    }

    private boolean empty() {
        return this.elemNumber == 0;
    }

    private boolean filled() {
        return this.elemNumber == this.capacity;
    }

    public void publish(int value) {
        if (this.filled()) {
            System.out.println("The array is already filled");
            return;
        }

        if (this.elemNumber < this.capacity) {
            if (this.last == this.capacity - 1) {
                this.last = -1;
            }
            this.last += 1;
            this.elemNumber += 1;
            this.values[this.last] = value;
        }
    }

    public void consume() {
        if (this.empty()) {
            System.out.println("The array is empty");
            return;
        }

        if (this.first == this.capacity - 1) {
            this.first = 0;
        }
        this.first += 1;
        this.elemNumber -= 1;
    }

    public static void main(String[] args) {
        CircularQueue cq = new CircularQueue(5);
        cq.publish(1);
        cq.publish(2);
        cq.publish(3);
        cq.publish(4);
        cq.publish(5);
        cq.consume();
        cq.consume();
        cq.consume();
        cq.publish(6);
        cq.publish(7);
        cq.publish(8);
        cq.consume();
        cq.consume();
        cq.publish(9);
        cq.publish(10);
        cq.consume();
        cq.publish(11);
    }
}

File: data-structure/python/queues/circular-queue.py
import numpy as np

class Queue:
    def __init__(self, capacity):
        self.__capacity = capacity
        self.__values = np.empty(capacity, dtype=int)
        self.__first = 0
        self.__last = -1
        self.__elem_number = 0

    def __empty(self):
        if self.__elem_number == 0:
            print('A fila está vazia')
            return True
        
    def __filled(self):
        if self.__elem_number == self.__capacity:
            print('A fila está cheia')
            return True
        
    def add(self, value):
        if self.__filled():
            return -1

        if self.__elem_number < self.__capacity:
                if(self.__last == self.__capacity -1):
                    self.__last = -1
                self.__last += 1
                self.__values[self.__last] = value
                self.__elem_number +=1
    
    def remove(self):
        if self.__empty():
            return -1
        
        self.__first += 1

        if self.__first == self.__capacity:
            self.__first = 0        

        self.__elem_number -= 1

    def first(self):
        return self.__first
    
    def last(self):
        return self.__last
        


queue = Queue(5)
queue.add(3)
queue.add(5)
queue.add(1)
queue.add(4)
queue.add(7)
print(queue.first())
print(queue.last())
queue.remove()
queue.remove()
queue.add(5)
queue.add(1)
queue.remove()
queue.remove()
queue.remove()
queue.remove()
queue.remove()
queue.add(1)
print(queue.first())
print(queue.last())
